src/device/usbd.c

```diff
@@ -450,7 +450,7 @@ void tud_task (void)
 // Helper to invoke class driver control request handler
 static bool invoke_class_control(uint8_t rhport, uint8_t drvid, tusb_control_request_t const * request)
 {
-  TU_ASSERT(_usbd_driver[drvid].control_request);
+  TU_VERIFY(_usbd_driver[drvid].control_request);
```

Comment:
we could actually remove this, the control request is mandatory for all driver, drvid is already valid to this point.

Reply:
I didn't realize it was mandatory. Removing the check seems reasonable, but maybe also adding a comment to state that the sof() function is optional in the struct definition.

Or maybe we can add a TU_ASSERT for all the function members in tud_init where they would be optimised out by the compiler...

Reply:
@pigrew yeah, you are right, since the driver array is const, even this will surely be optimized by compiler. Maybe just leave it this for now. I am still not sure what to do with SOF. It is probably required for ISO transfer, but yes, it is optional for most class driver.

Reply:
But again, I am not sure if this would change :D. Maybe there is a class that won't bother to do the control request.

Reply:
Currently the MIDI class doesn't bother, it has empty function definitions. But, the MIDI spec does define a pair of requests.

Reply:
It is still better to have empty function than leaving it NULL I think. Yeah, let's TU_ASSERT() all of them (except for the SOF) when init the stack :+1: :+1: